### PR TITLE
rail-fence-cipher: add to track

### PR DESCRIPTION
**Pull request by LeonDsouza**
---

Reviewer Resources:

[Track Policies](https://github.com/exercism/xjava/blob/master/POLICIES.md#event-checklist)

### CODE REVIEW

**Comment by stkent** [comment]
Oh, now I'm confused; which PR is the most up-to-date, @LeonDsouza? I see this one includes those config changes I mentioned in your other PR.

**Comment by Smarticles101** [comment]
In order to fix the failing build, you need to add a folder src/main/java and put an empty .keep file in it in order for git to see it let you commit that folder. Otherwise, the build fails because the folder doesn't exist. 

Also, as a side note, it's perfectly fine to keep using the original pull request when you push new changes. Github adds any more pushed changes to the pull request, and it also allows you to edit the title when you like. Just using one pull request helps to keep out confusion :)

File: config.json [hunk 1]
```diff
@@ -13,6 +13,15 @@
 
       ]
     },
+{
+      "slug": "rail-fence-cipher",
+      "core": true,
+      "unlocked_by": null,
+      "difficulty": 1,
+      "topics": [
```
Comment (Smarticles101) [inline config.json]: Would you be able to estimate a difficulty and add topics? If not, we can open a separate issue for that once this gets merged.

File: exercises/rail-fence-cipher/src/test/java/RailFenceCipherTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	public void encodeWithTwoRails() {
+		railFenceCipher = new RailFenceCipher(2);
+		Assert.assertEquals("XXXXXXXXXOOOOOOOOO", railFenceCipher.getEncryptedData("XOXOXOXOXOXOXOXOXO"));
+	}
+
+	@Ignore("Remove to run test")
+	@Test
+	public void encodeWithThreeRails() {
+		railFenceCipher = new RailFenceCipher(3);
+		Assert.assertEquals("WECRLTEERDSOEEFEAOCAIVDEN", railFenceCipher.getEncryptedData("WEAREDISCOVEREDFLEEATONCE"));
```
Comment (Smarticles101) [inline exercises/rail-fence-cipher/src/test/java/RailFenceCipherTest.java]: These lines might be easier to read with each parameter on a different line.

File: exercises/rail-fence-cipher/src/example/java/RailFenceCipher.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            for (int j = 0; j < railFence.get(i).size(); ++j) {
+                buffer.set(railFence.get(i).get(j), cipherText.charAt(counter));
+                counter++;
+            }
+        }
+
+        decrypted = buffer.toString().replaceAll(",", "").replaceAll("\\s+", "").replace("[", "").replace("]", "");
+        return decrypted;
+    }
+}
```
Comment (Smarticles101) [inline exercises/rail-fence-cipher/src/example/java/RailFenceCipher.java]: I don't have any comments on the example implementation. @FridaTveit usually does if she cares to look at it

File: config.json (hunk 1 above)
Comment (FridaTveit) [reply config.json]: Also `core` should be false. We've already decided a set of core exercises and we probably don't want to add to them at this point. See [this issue](https://github.com/exercism/java/issues/707) for our discussion about which exercises should be core exercises. Also [this issue](https://github.com/exercism/java/issues/758) and [this PR](https://github.com/exercism/java/pull/760) where we changed the `config.json` structure after finding out exercises should only be unlocked by core exercises.

If you want @LeonDsouza, you could look through the core exercises (listed in [this PR](https://github.com/exercism/java/pull/760)) and choose want you think should unlock this one. But it can also be kept as `"core": false` and `"unlocked_by": null` :)

Let me know if any of that didn't make sense! :)

File: exercises/rail-fence-cipher/src/example/java/RailFenceCipher.java (hunk 3 above)
Comment (FridaTveit) [reply exercises/rail-fence-cipher/src/example/java/RailFenceCipher.java]: The most important thing is that this file should be in `.meta/src/reference/java/` instead of `src/example/java/`. Sorry about that, it's been changed very recently :) You can look at the other exercises @LeonDsouza to make sure your folder structure matches theirs :)

File: config.json (hunk 1 above)
Comment (FridaTveit) [reply config.json]: It also needs a `uuid`. These were introduced quite recently, see [this PR](https://github.com/exercism/java/pull/704). You can use [this tool](https://www.uuidgenerator.net/) to generate one :)

Comment (FridaTveit) [reply config.json]: I just found [this document](https://github.com/exercism/docs/blob/master/language-tracks/configuration/exercises.md) which explains very well what needs to be in this file (including `uuid`, `unlocked_by` and `core`) :)

**Comment by LeonDsouza** [comment]
Did those changes @Smarticles101 @stkent @FridaTveit. Got the build to succeed! 👍

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        String buffer = "";
+        for (String s : railFence) {
+            buffer = buffer.concat(s);
+        }
+        return buffer;
+    }
+
+    String getDecryptedData(String cipherText) {
+        String decrypted;
+        ArrayList<ArrayList<Integer>> railFence = new ArrayList<>(noOfRails);
```
Comment (FridaTveit) [inline exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: Is there any particular reason you're storing each rail as a list of integers? You could just store it as a string and append characters to it as you go through the ciphertext. That should make it easier to get the final output at the end :)

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        for (int i = 0; i < noOfRails; ++i) {
+            for (int j = 0; j < railFence.get(i).size(); ++j) {
+                buffer.set(railFence.get(i).get(j), cipherText.charAt(counter));
+                counter++;
+            }
+        }
+
+        decrypted = buffer.toString().replaceAll(",", "").replaceAll("\\s+", "").replace("[", "").replace("]", "");
```
Comment (FridaTveit) [inline exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: Do you really need these replaces?

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        String decrypted;
+        ArrayList<ArrayList<Integer>> railFence = new ArrayList<>(noOfRails);
+
+        for (int i = 0; i < noOfRails; ++i) {
+            railFence.add(new ArrayList<>(cipherText.length()));
+        }
+
+        int rowIncrementFactor = 0, step = 1;
+
+        for (int i = 0; i < cipherText.length(); ++i) {
```
Comment (FridaTveit) [inline exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: Again, this is a bit hard to follow. Say you have n rails, then you want to go through the ciphertext and add each letter to the correct rail. You can find the correct rail by `i % n`. If `i % n` is 0 then you add it to the first rail, if `i % n` is 1 then you add it to the second rail. Does that make sense?

**Comment by FridaTveit** [comment]
Hi @LeonDsouza :) How's it going with this? :)

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        this.noOfRails = noOfRails;
+    }
+
+    String getEncryptedData(String plainText) {
+        List<String> railFence = new ArrayList<>();
+        for (int i = 0; i < noOfRails; i++) {
+            railFence.add("");
+        }
+
+        int number = 0, increment = 1;
```
[the remark this answers is not in the document]
Comment (LeonDsouza) [reply exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: I'm actually using the increment variable to know the extent of how far down or up i can go

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java (hunk 4 above)
Comment (LeonDsouza) [reply exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: I've stored it in a kind of matrix form for a 2 loop iteration

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (FridaTveit) [reply exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: Okay. That still doesn't explain why it's called `increment`. The name of the variable should explain what you're using it for :)

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java (hunk 4 above)
Comment (FridaTveit) [reply exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: Sure, but since you're dealing with characters it makes more sense to store it as characters rather than integers. You can have a list of strings and loop through the characters of each string.

And as a tip for the future, it would be really helpful if you could respond to reviews relatively soon after we do them. For big reviews like this it takes time for me to get into the mindset of the code and understand what you're doing. If you had made these comments a month ago when I first reviewed this code I would have been able to give way more useful answers. And I'm guessing it would be easier for you too to change your code right after you wrote it instead of more than a month after.
I know you're probably very busy with work and things, but in that case it might be easier for you to work on smaller issues rather than trying to implement a whole exercise? Like I said, this is just a suggestion for the future since PRs that remain open for months become really hard to review. But I do really appreciate you trying to help improve the track :)

**Comment by stkent** [comment]
Poke @LeonDsouza :) How's it going? If you're too busy to finish this one off, let us know and we'll carry it across the finish line. I wouldn't want your hard work so far to be wasted 😢

**Comment by sjwarner-bp** [comment]
Seeing as this has been inactive a little while, does anyone mind if I pick it up and finish it off so this work isn't wasted? :slightly_smiling_face: 

As a sidenote, what normally happens with abandoned PRs?

**Comment by stkent** [comment]
> Seeing as this has been inactive a little while, does anyone mind if I pick it up and finish it off so this work isn't wasted? :slightly_smiling_face: 

I think it's safe to do so at this point.

> As a sidenote, what normally happens with abandoned PRs?

I'm not sure this has happened often enough for us to have a 'normal' response. We trialed running probot on the repository a while back, but that was mostly for support-related issues and it was missing features we needed.

**Comment by sjwarner-bp** [comment]
@stkent Thanks for letting me know!

As for finishing this one off, I've never amended someone else's PR before. I've seen before (example #1060) where others may push to a PR they haven't authored. If anyone could let me know how this is possible I'd be really grateful :slightly_smiling_face:

**Comment by Smarticles101** [comment]
@sjwarner-bp It's as simple as cloning the branch that will be merged through the pull request and committing to it! :)

**Comment by sjwarner-bp** [comment]
Ok, so we have a passing build (I've updated `config.json`) - but I still think that the reference solution needs a little work :slightly_smiling_face:

File: config.json [hunk 8]
```diff
@@ -483,6 +483,18 @@
...
+    {
+      "core": false,
+      "difficulty": 6,
+      "slug": "rail-fence-cipher",
+      "topics": [
+        "strings",
+        "loops",
+        "conditionals"
+      ],
+      "unlocked_by": "atbash-cipher",
```
Comment (FridaTveit) [inline config.json]: `atbash-cipher` isn't a core exercise so it can't be unlocked by that. `rotational-cipher` would probably be a good exercise for it to be unlocked by? :)

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
+public class RailFenceCipher {
```
Comment (FridaTveit) [inline exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: This could be package private :)

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    private static int key;
```
Comment (FridaTveit) [inline exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: I don't think this or any of the methods should be static. Either all the methods are static and you pass in the key to each method, or the methods are all instance methods and you pass in the key to the constructor. Since we prefer instance methods according to our [policies doc](https://github.com/exercism/java/blob/master/POLICIES.md#prefer-instance-methods) I would suggest the latter :)

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+    public RailFenceCipher(int key) {
+        RailFenceCipher.key = key;
+    }
+
+    public static String getEncryptedData(String message) {
+        String[] lines = splitIntoLines(message, false);
+        String result = "";
```
Comment (FridaTveit) [inline exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: You could use a StringBuilder :)

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        int charCount = 0;
+        for (int i = 0; i < key; ++i) {
+            while (lines[i].contains("?")) {
+                String letter = String.valueOf(message.charAt(charCount));
+                lines[i] = lines[i].replaceFirst("\\?", letter);
+                charCount++;
+            }
+        }
+
+        String result = "";
```
Comment (FridaTveit) [inline exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: Could use StringBuilder :)

File: exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            String letter = String.valueOf(c);
+            result[lineCount] += encrypted ? "?" : letter;
+            direction *= lineCount == 0 || lineCount == key - 1 ? -1 : 1;
+            lineCount += direction;
+        }
+        return result;
+    }
+
+    private static String[] generateEmptyStrings(int num) {
+        String[] strings = new String[num];
```
Comment (FridaTveit) [inline exercises/rail-fence-cipher/.meta/src/reference/java/RailFenceCipher.java]: Could you do:
```
String[] strings = new String[num];
Arrays.fill(strings, "");
return strings;
```
? :)

**Comment by sjwarner-bp** [comment]
Oops! Silly mistakes - hopefully they're all fixed up now :slightly_smiling_face:
